File: src/sync/test_helpers.rs
use super::*;
use crate::dag;
use crate::db;
use crate::db::test_helpers::*;
use crate::db::{Commit, Whence};
use crate::util::rlog;
use str_macro::str;

// See db::test_helpers for add_local, add_snapshot, etc. We can't put add_local_rebase
// there because sync depends on db, and add_local_rebase depends on sync.

// add_sync_snapshot adds a sync snapshot off of the main chain's base snapshot and
// returns it (in chain order). Caller needs to supply which commit to take indexes
// from because it is context dependent (they should come from the parent of the
// first commit to rebase, or from head if no commits will be rebased).
pub async fn add_sync_snapshot<'a>(
    chain: &'a mut Chain,
    store: &dag::Store,
    take_indexes_from: usize,
    lc: rlog::LogContext,
) -> Chain {
    assert!(chain.len() >= 2); // Have to have at least a genesis and a local commit on main chain.

    let mut maybe_base_snapshot: Option<&Commit> = None;
    for i in (1..chain.len()).rev() {
        if chain[i - 1].meta().is_snapshot() {
            maybe_base_snapshot = Some(&chain[i - 1]);
            break;
        }
    }
    if maybe_base_snapshot.is_none() {
        panic!("main chain doesnt have a snapshot or local commit");
    }
    let base_snapshot = maybe_base_snapshot.unwrap();
    let mut sync_chain: Chain = vec![];

    // Add sync snapshot.
    let ssid = format!("sync_server_state_id_{}", chain.len());
    let indexes = db::read_indexes(&chain[take_indexes_from]);
    let w = db::Write::new_snapshot(
        Whence::Hash(base_snapshot.chunk().hash().to_string()),
        base_snapshot.mutation_id(),
        ssid,
        store.write(lc.clone()).await.unwrap(),
        indexes,
    )
    .await
    .unwrap();
    w.commit(SYNC_HEAD_NAME).await.unwrap();
    let (_, commit, _) = db::read_commit(
        Whence::Head(str!(SYNC_HEAD_NAME)),
        &store.read(lc.clone()).await.unwrap().read(),
    )
    .await
    .unwrap();
    sync_chain.push(commit);

    sync_chain
}

File: src/sync/test-helpers.ts
import {expect} from '@esm-bundle/chai';
import type {Chain} from '../db/test-helpers';
import type * as dag from '../dag/mod';
import * as db from '../db/mod';
import * as sync from '../sync/mod';

// See db.test_helpers for addLocal, addSnapshot, etc. We can't put addLocalRebase
// there because sync depends on db, and addLocalRebase depends on sync.

// addSyncSnapshot adds a sync snapshot off of the main chain's base snapshot and
// returns it (in chain order). Caller needs to supply which commit to take indexes
// from because it is context dependent (they should come from the parent of the
// first commit to rebase, or from head if no commits will be rebased).

export async function addSyncSnapshot(
  chain: Chain,
  store: dag.Store,
  takeIndexesFrom: number,
): Promise<Chain> {
  expect(chain.length >= 2).to.be.true;

  let maybeBaseSnapshot: db.Commit | undefined;
  for (let i = chain.length - 1; i > 0; i--) {
    if (chain[i - 1].meta().isSnapshot()) {
      maybeBaseSnapshot = chain[i - 1];
      break;
    }
  }
  if (maybeBaseSnapshot === undefined) {
    throw new Error('main chain doesnt have a snapshot or local commit');
  }
  const baseSnapshot = maybeBaseSnapshot;
  const syncChain: Chain = [];

  // Add sync snapshot.
  const cookie = `sync_cookie_${chain.length}`;
  const indexes = db.readIndexes(chain[takeIndexesFrom]);
  await store.withWrite(async dagWrite => {
    const w = await db.Write.newSnapshot(
      db.whenceHash(baseSnapshot.chunk.hash),
      baseSnapshot.mutationID(),
      cookie,
      dagWrite,
      indexes,
    );
    await w.commit(sync.SYNC_HEAD_NAME);
  });
  const [, commit] = await store.withRead(async dagRead => {
    return await db.readCommit(db.whenceHead(sync.SYNC_HEAD_NAME), dagRead);
  });
  syncChain.push(commit);

  return syncChain;
}
